cli/flox-rust-sdk/src/models/manifest.rs

```diff
@@ -890,20 +895,96 @@ pub struct PackageInsertion {
     pub already_installed: HashMap<String, bool>,
 }
 
-/// A package to install.
+/// Any kind of package that can be installed via `flox install`.
+#[derive(Debug, Clone, PartialEq)]
+pub enum PackageToInstall {
+    Catalog(CatalogPackage),
+    Flake(FlakePackage),
+}
+
+impl PackageToInstall {
+    pub fn id(&self) -> &str {
+        match self {
+            PackageToInstall::Catalog(pkg) => &pkg.id,
+            PackageToInstall::Flake(pkg) => &pkg.id,
+        }
+    }
+
+    pub fn set_id(&mut self, id: impl AsRef<str>) {
+        let id = String::from(id.as_ref());
+        match self {
+            PackageToInstall::Catalog(pkg) => pkg.id = id,
+            PackageToInstall::Flake(pkg) => pkg.id = id,
+        }
+    }
+}
+
+impl FromStr for PackageToInstall {
+    type Err = ManifestError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match Url::parse(s) {
+            Ok(url) => {
+                let id = infer_flake_install_id(&url)?;
+                Ok(PackageToInstall::Flake(FlakePackage { id, url }))
+            },
+            _ => Ok(PackageToInstall::Catalog(s.parse()?)),
+        }
+    }
+}
+
+/// Tries to infer an install id from the flake ref URL, or falls back to "flake".
+fn infer_flake_install_id(url: &Url) -> Result<String, ManifestError> {
+    if url.scheme() == "github" {
+        url.path()
+            .split('/')
+            .nth(1)
+            .map(|s| s.to_string())
+            .ok_or(ManifestError::InvalidFlakeRef(url.to_string()))
+    } else if url.scheme() == "https" && url.domain().is_some_and(|host| host == "github.com") {
+        let mut segments = url
+            .path_segments()
+            .ok_or(ManifestError::InvalidFlakeRef(url.to_string()))?;
+        segments
+            .nth(1)
+            .map(|s| s.to_string())
+            .ok_or(ManifestError::InvalidFlakeRef(url.to_string()))
+    } else {
+        Ok("flake".to_string())
+    }
+}
```

Comment:
suggestion: try use fragment/attrpath if available, dont special case github (we discussed that bad even bad cases of applied heuristics like

```
github:flox/flox/cae31a966476e42293f2bb9b085fa2b0d86896e7b5
```

would be ok.


with reusing the install id extraction from catalog installbles a possible diff is:


```diff
diff --git a/cli/flox-rust-sdk/src/models/manifest.rs b/cli/flox-rust-sdk/src/models/manifest.rs
index 497e801c2..b2ef9232a 100644
--- a/cli/flox-rust-sdk/src/models/manifest.rs
+++ b/cli/flox-rust-sdk/src/models/manifest.rs
@@ -935,23 +935,25 @@ impl FromStr for PackageToInstall {
 
 /// Tries to infer an install id from the flake ref URL, or falls back to "flake".
 fn infer_flake_install_id(url: &Url) -> Result<String, ManifestError> {
-    if url.scheme() == "github" {
-        url.path()
-            .split('/')
-            .nth(1)
-            .map(|s| s.to_string())
-            .ok_or(ManifestError::InvalidFlakeRef(url.to_string()))
-    } else if url.scheme() == "https" && url.domain().is_some_and(|host| host == "github.com") {
-        let mut segments = url
-            .path_segments()
-            .ok_or(ManifestError::InvalidFlakeRef(url.to_string()))?;
-        segments
-            .nth(1)
-            .map(|s| s.to_string())
-            .ok_or(ManifestError::InvalidFlakeRef(url.to_string()))
-    } else {
-        Ok("flake".to_string())
+    if let Some(fragment) = url.fragment() {
+        let attr_path = fragment
+            .rsplit_once('^')
+            .map(|(attr_path, _)| attr_path)
+            .unwrap_or(fragment);
+        let install_id = install_id_from_attr_path(attr_path, url.as_ref())?;
+        if !install_id.is_empty() {
+            return Ok(install_id);
+        }
     }
+
+    // use path because `github:` and co. are `cannot-be-a-base` urls
+    // for which "path-segments" are undefined.
+    // `Url::path_segments` will return `None` for such urls.
+    url.path()
+        .split('/')
+        .last()
+        .map(|s| s.to_string())
+        .ok_or(ManifestError::InvalidFlakeRef(url.to_string()))
 }
 
 /// Extracts only the catalog packages from a list of packages to install.
@@ -1022,43 +1024,7 @@ impl FromStr for CatalogPackage {
             None => (descriptor.to_string(), None),
         };
 
-        let install_id = {
-            let mut install_id = None;
-            let mut cur = String::new();
-
-            let mut start_quote = None;
-
-            for (n, c) in attr_path.chars().enumerate() {
-                match c {
-                    '.' if start_quote.is_none() => {
-                        let _ = install_id.insert(std::mem::take(&mut cur));
-                    },
-                    '"' if start_quote.is_some() => start_quote = None,
-                    '"' if start_quote.is_none() => start_quote = Some(n),
-                    other => cur.push(other),
-                }
-            }
-
-            if start_quote.is_some() {
-                return Err(ManifestError::MalformedStringDescriptor {
-                    msg: "unclosed quote".to_string(),
-                    desc: descriptor.to_string(),
-                });
-            }
-
-            if !cur.is_empty() {
-                let _ = install_id.insert(cur);
-            }
-
-            if install_id.is_none() {
-                return Err(ManifestError::MalformedStringDescriptor {
-                    msg: "attribute path is empty".to_string(),
-                    desc: descriptor.to_string(),
-                });
-            }
-
-            install_id.unwrap()
-        };
+        let install_id = install_id_from_attr_path(&attr_path, descriptor)?;
 
         Ok(Self {
             id: install_id,
@@ -1068,6 +1034,45 @@ impl FromStr for CatalogPackage {
     }
 }
 
+/// parses a dot-separated attribute path into
+fn install_id_from_attr_path(attr_path: &str, descriptor: &str) -> Result<String, ManifestError> {
+    let mut install_id = None;
+    let mut cur = String::new();
+
+    let mut start_quote = None;
+
+    for (n, c) in attr_path.chars().enumerate() {
+        match c {
+            '.' if start_quote.is_none() => {
+                let _ = install_id.insert(std::mem::take(&mut cur));
+            },
+            '"' if start_quote.is_some() => start_quote = None,
+            '"' if start_quote.is_none() => start_quote = Some(n),
+            other => cur.push(other),
+        }
+    }
+
+    if start_quote.is_some() {
+        return Err(ManifestError::MalformedStringDescriptor {
+            msg: "unclosed quote".to_string(),
+            desc: descriptor.to_string(),
+        });
+    }
+
+    if !cur.is_empty() {
+        let _ = install_id.insert(cur);
+    }
+
+    if install_id.is_none() {
+        return Err(ManifestError::MalformedStringDescriptor {
+            msg: "attribute path is empty".to_string(),
+            desc: descriptor.to_string(),
+        });
+    }
+
+    Ok(install_id.unwrap())
+}
+
 impl From<&CatalogPackage> for Vec<(&'static str, String)> {
     fn from(val: &CatalogPackage) -> Self {
         let mut vec = vec![("pkg-path", val.pkg_path.clone())];

```